flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java

```diff
@@ -259,7 +259,7 @@
 
 	private PartitionReleaseStrategy partitionReleaseStrategy;
 
-	private SchedulingTopology schedulingTopology;
+	private SchedulingTopology<?, ?> schedulingTopology;
```

Comment:
Do you mean to avoid the case that a lazy attaching did not happen?
If so, I'm thinking whether we can make vertex building in ExecutionGraph ctor and drop the lazy attaching mechanism. In production, the vertex attaching is only invoked right after the EG is created in ExecutionGraphBuilder. And attaching vertices after a job starts scheduling seems to be an undefined behavior currently which would not work properly.

Reply:
It would mainly be a safeguard against using this class in a wrong way or against faulty refactorings. What you are mentioning goes a couple of steps farther. At the moment I think we don't use the lazy attaching mechanism and, hence, we could remove it.

Reply:
Shall we open a ticket to remove the lazy attaching mechanism, and build vertices and do `executionTopology`/`partitionReleaseStrategy` initialization in `ExecutionGraph` constructor?
In this way, `executionTopology` can be final, so that we do not need a `FailingSchedulingTopology` to ensure its assigned correctly.

Reply:
Yes let's create a ticket for it. But it is not super critical to do.

Reply:
FLINK-14510 is opened to track it. Yet we can do it later.